packages/contracts/contracts/appDefinitions/HighRollerApp.sol

```diff
@@ -0,0 +1,179 @@
+pragma solidity 0.4.25;
+pragma experimental "ABIEncoderV2";
+
+import "../lib/Transfer.sol";
+
+/*
+  This contract is for a dice rolling game.
+  Two players roll two dice each.
+  The winner is the player whose sum of dice outcomes is highest.
+*/
+contract HighRollerApp {
+
+  enum ActionType {
+    COMMIT_TO_HASH,
+    COMMIT_TO_NUM
+  }
+
+  enum Stage {
+    COMMITTING_HASH, 
+    COMMITTING_NUM, 
+    DONE
+  }
+
+  enum Player {
+    FIRST,
+    SECOND
+  }
+
+  struct AppState {
+    address[2] playerAddrs;
+    Stage stage;
+    bytes32 salt;
+    bytes32 commitHash;
+    uint256 commitNum;
+    uint256 hashNum;
+  }
+
+  struct Action {
+    ActionType actionType;
+    uint256 number;
+    bytes32 actionHash;
+  }
+
+  function isStateTerminal(AppState state)
+    public
+    pure
+    returns (bool)
+  {
+    return state.stage == Stage.DONE;
+  }
+
+  function getTurnTaker(AppState state)
+    public
+    pure
+    returns (Player)
+  {
+    if (state.stage == Stage.COMMITTING_NUM) {
+      return Player.SECOND;
+    }
+    return Player.FIRST;
+  }
+
+  function applyAction(AppState state, Action action)
+    public
+    pure
+    returns (bytes)
+  {
+    AppState memory nextState = state;
+    if (action.actionType == ActionType.COMMIT_TO_HASH) {
+      require(state.stage == Stage.COMMITTING_HASH, "Cannot apply COMMIT_TO_HASH on COMMITTING_HASH");
+      nextState.stage = Stage.COMMITTING_NUM;
+
+      nextState.commitHash = action.actionHash;
+    } else if (action.actionType == ActionType.COMMIT_TO_NUM) {
+      require(state.stage == Stage.COMMITTING_NUM, "Cannot apply COMMITTING_NUM on COMMITTING_NUM");
+      nextState.stage = Stage.DONE;
+
+      nextState.commitNum = action.number;
+    } else {
+      revert("Invalid action type");
+    }
+    return abi.encode(nextState);
+  }
+
+  function resolve(AppState state, Transfer.Terms terms)
+    public
+    pure
+    returns (Transfer.Transaction)
+  {
+    bytes32 salt = state.salt;
+    uint256 playerFirstNumber = state.hashNum;
+    uint256 playerSecondNumber = state.commitNum;
```

Comment:
Adding new variables is actually costly in Solidity; I think you could probably just rename the variables in the `AppState` struct to be more descriptive versus needing to do this declaration of a new variable.

Reply:
e.g., `state.playerSecondNumber` or something